File: DSA/MouliRathor_590019184/Day_3/Question2.java
class Solution {
    public static int firstOccurrence(int[] arr, int target) {
        int start = 0, end = arr.length - 1;
        int result = -1;
        
        while (start <= end) {
            int mid = start + (end - start) / 2;
            
            if (arr[mid] == target) {
                result = mid;
                end = mid - 1; 
            } else if (arr[mid] < target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        
        return result;
    }
    public static int lastOccurrence(int[] arr, int target) {
        int start = 0, end = arr.length - 1;
        int result = -1;
        
        while (start <= end) {
            int mid = start + (end - start) / 2;
            
            if (arr[mid] == target) {
                result = mid;
                start = mid + 1;
            } else if (arr[mid] < target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        
        return result;
    }

    public static int countFreq(int[] arr, int target) {
        int first = firstOccurrence(arr, target);
        int last = lastOccurrence(arr, target);
        if (first == -1 || last == -1) {
            return 0;
        }
        return last - first + 1;
    }
    
    public static void main(String[] args) {
        int[] arr = {1, 1, 2, 2, 2, 2, 3};
        int target = 2;
        System.out.println(countFreq(arr, target)); // Output: 4
    }
}

File: DSA/Abhishek_590028674/Day_3/Question2.c
#include <stdio.h>

int bSearch(int arr[], int arrSize, int target, int boolFirst){
    int l = 0, h = arrSize-1, result = -1;
    while(l <= h){
        int m = l + (h-l)/2;
        if(arr[m] == target){
            result = m;
            if(boolFirst) 
                h = m-1;
            else
                l = m+1;
        } 
        else if(arr[m] < target)
            l = m+1;
        else
            h = m-1;
    }
    return result;
}

void targetCount(int arr[], int arrSize, int target){
    int first = bSearch(arr, arrSize, target, 1);
    if(first == -1){
        printf("0");
        return;
    }
    int last = bSearch(arr, arrSize, target, 0);
    printf("%d", last-first+1);
}

int main(){
    int exampleArr[] = {1, 1, 2, 2, 2, 3, 3};
    int exampleTarget = 2;
    targetCount(exampleArr, sizeof(exampleArr)/sizeof(exampleArr[0]), exampleTarget);
    return 0;
}